### PR TITLE
[jest] add type inference

### PR DESCRIPTION
**Pull request by antoinebrault**
Updated jest typings to use latest typescript features to infer mock types. 

### Breaking Change

With types inferred, mockImplementation(), mockReturnValue(), mock...() will not compile if they are not used with the right types.

----

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: **private project**
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/jest/index.d.ts [hunk 1]
```diff
@@ -762,12 +763,12 @@ declare namespace jest {
         new (...args: any[]): any;
     }
 
-    interface Mock<T = {}> extends Function, MockInstance<T> {
-        new (...args: any[]): T;
-        (...args: any[]): any;
+    interface Mock<T, Y extends any[]> extends Function, MockInstance<T, Y> {
```
Comment (jwbay) [inline types/jest/index.d.ts]: Please restore the default type parameter and add one for Y. I don't think making these required is worth the breaking change, and trying to "force people" to use stronger types in tests is very opinionated. A large percentage of `jest.Mock` usages don't touch values, only call counts. Also keep in mind the inference for stronger mock types will only help with spyOn. Every other declaration will need to have the parameters passed explicitly, which is going to be a bit painful since Y is an arguments tuple that needs to be extracted.

File: types/jest/index.d.ts [hunk 2]
```diff
@@ -939,26 +940,22 @@ declare namespace jest {
...
     /**
      * Represents the result of a single call to a mock function.
      */
     interface MockResult {
+        type: 'return' | 'throw' | 'incomplete';
         /**
-         * True if the function threw.
-         * False if the function returned.
-         */
-        isThrow: boolean;
```
Comment (jwbay) [inline types/jest/index.d.ts]: Context for this change? Was this property removed from Jest?

I see isThrow still in the docs, and no 'type' at all: https://jestjs.io/docs/en/mock-function-api#mockfnmockresults

File: types/jest/index.d.ts [hunk 3]
```diff
@@ -198,7 +197,9 @@ declare namespace jest {
      *   spy.mockRestore();
      * });
      */
-    function spyOn<T extends {}, M extends keyof T>(object: T, method: M, accessType?: 'get' | 'set'): SpyInstance<T[M]>;
+    function spyOn<T extends {}, M extends keyof T>(object: T, method: M, accessType: 'get'): SpyInstance<T[M], []>;
+    function spyOn<T extends {}, M extends keyof T>(object: T, method: M, accessType: 'set'): SpyInstance<void, [T[M]]>;
```
Comment (jwbay) [inline types/jest/index.d.ts]: Since `T[M]` is the function type here, should this be `SpyInstance<void, ExtractArgs<T[M]>>`?

**Review by jwbay — CHANGES_REQUESTED** [review]

File: types/jest/index.d.ts (hunk 3 above)
Comment (antoinebrault) [reply types/jest/index.d.ts]: No, TypeScript considers a setter/getter to be a property, so the type of the property is used here. It is not a function.

File: types/jest/index.d.ts (hunk 1 above)
Comment (nasreddineskandrani) [reply types/jest/index.d.ts]: Please we want to move away from `any` when we are with typescript => so no default value with `any`

Arguments:
1.
We need this contract for robust integrated test. I think this had to be corrected long ago (ts 2.8)
in other case the test are reall weak.


2.
This fix will make jasmine behind compared to jest and will raise the adoption of the library.

3.
https://palantir.github.io/tslint/rules/no-any/
`Using any as a type declaration nullifies the compile-time benefits of the type system.`

@rickhanlonii can we have your opinion here, please?

Comment (rickhanlonii) [reply types/jest/index.d.ts]: > A large percentage of jest.Mock usages don't touch values, only call counts

@jwbay can you explain this with an example of what you can do now and what this change would require?

Comment (jwbay) [reply types/jest/index.d.ts]: > I know that is it very opinionated, but shouldn't we, as a community, enforce best practices?

To a point. We need to balance the benefits and costs considering the pain of breaking changes. As a data point, it's possible to make a package so annoyingly type-safe the change is reverted: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/26813

> I strongly believe that testing should be done the safest way possible, and "any" type should be avoided at all cost.

You are absolutely entitled to that strong opinion in your projects. The issue here is that this change impacts a large number of developers across the world, from enterprise teams to personal projects (this package gets 2 million downloads per week), for value that people will question. This change as-is isn't really just opinionated—it's prescriptive.

> I don't agree that stronger mock types will only help for spyOn. There is a couple of example I changed in jest-tests (mockContextVoid, mockContextString, spy3Mock, spy4) where the typing of arguments would be lost and potentially mislead people.

That's fair, I didn't qualify my statement correctly. spyOn is the only way to infer types *from existing functions*. Everything else will either need explicit types, a runtime function to extract them via inference, or some kind of conditional type to convert `typeof MyFunction` to a correctly populated `jest.Mock`. We should consider adding that last option with this change, actually, assuming it goes forward. Looks like it's just `T extends (...args: infer R) => infer V ? jest.Mock<V, R> : never`

> If you don't care about tuple generics, you can still use <any, any[]> or without typing :

That's part of the problem with this. The <any, any[]> is often just noise. Reading noise sucks and writing noise really sucks. Here are a couple examples:

```tsx
test('foo', () => {
  someFunction();
  // type params are just noise here
  (foo as jest.Mock<any, any[]>).mockClear();
  someFunction();
  expect(fooMock).toHaveBeenCalledTimes(1);
});

test('foo', () => {
  // more noise
  (foo as jest.Mock<any, any[]>).mockName('MyMock');
  const rendered = render(<Component foo={foo} />);
  expect(rendered).toMatchSnapshot();
});
```

I think providing defaults for the cases where the type parameters truly don't matter is a reasonable ask. How about a compromise of defaulting to `never`? That way you can't interact with arguments or return values unsafely by default, and you can't supply an invalid implementation. It would still be a massive breaking change, but at least the types involved are being touched in some way, so it's more reasonable as a value proposition.

I'd really like to get the thoughts of other maintainers here considering the impact: @NoHomey, @asvetliakov, @alexjoverm, @epicallan, @ikatyang, @wsmd, @JamieMason, @douglasduteil, @ahnpnl, @joshuakgoldberg, @UselessPickles, @r3nya, @hotell, @sebald, @andys8

As a note for impact, things like this are going to start popping up in codebases consuming this change:

```ts
// this function only touches one or two properties of a larger interface
mockFoo.mockReturnValue({
  x: 42
} as Partial<RealReturnType> as RealReturnType);

mockFoo.mockImplementation((a, _b, _c, _iReallyDontCare, _makeItStop) => {
  return Promise.resolve(a)
});
```

Comment (JoshuaKGoldberg) [reply types/jest/index.d.ts]: Sorry, it's a little hard to parse through all the discussion and changes - what exactly are the breaking changes? Where are folks now being forced to use types when things defaulted to `any` before?

It sounds like they might be significant _(for better or for worse)_, so even if they're the correct thing to do, there should be a clear summary of what's different. Jest is a popular enough library that any breaking change, no matter how small, is likely confusing and painful to some users.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]

File: types/jest/index.d.ts (hunk 1 above)
Comment (antoinebrault) [reply types/jest/index.d.ts]: updated description with breaking change

Comment (JoshuaKGoldberg) [reply types/jest/index.d.ts]: Thanks!

Although the rest of the changes look _excellent_ (really looking forward to being able to strictly type spies/mocks!), that one point of _requiring_ explicit type definitions on variables is not a good thing for many users. IMO we should default to `any` when not provided, not `never`.

Agreed that it's often better practice to explicitly type things, but not all users are interested in best practices, or the same best practices. Some users are just interested in having tests work out of the box, and any added pain given to them by TypeScript is a negative - even if it gives better type checking. A few examples:
* Large codebases converting from JavaScript would want as few places to _need_ to explicitly type variables as possible.
* In a lot of folks' minds, unit tests don't need nearly the same level of type checking as source code, and if you explicitly fail/error within a few milliseconds, needing to write & update them constantly just gets in the way.

Speaking more generally: what are "best practices"? If your goals are to have completely strict type safety, then sure, enforcing explicit types helps enforce your personal best practices. But not everybody has that same goal. Some just want to write code quickly, be able to define a few non-object constructs such as interfaces, and get some improved level of IDE features. IMO we shouldn't sacrifice one use case (stricter typing) for others (quicker development, easier TS migration).

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
-1 on the breaking change of jest.Mock and jest.SpyInstance needing to specify generics <return value, args type>. 

+1 on everything else.

File: types/jest/index.d.ts (hunk 1 above)
Comment (nasreddineskandrani) [reply types/jest/index.d.ts]: I understand all your arguments and i think it's case by case, team by team, project by project. A 
- Migration may require at the beginning an `any` default value. 
- Experienced people may not need to force them do it right when at the opposite other team needs it. 

So i am calling a crazy idea. 
How to make this opinionated default value **parametrizable ** for each of us depending on the needs.

Comment (nasreddineskandrani) [reply types/jest/index.d.ts]: ok! i am trying the flags to see if we can save the situation :) it by config
But i did what test that shows me a pure inconsistency
image

if the plan is to add only defaults value `any` for jest.Mock and jest.SpyInstance>
The compiler complaining for the second test but not the first one => `inconsistency`

Comment (nasreddineskandrani) [reply types/jest/index.d.ts]: `noImplicitAny` didn't work.  

 if default `any` used in jest type:
I found a way by wrapping the type
`type CustomSpyInstance<A = never> = jest.SpyInstance<A>;` which override `any`
with this it will prevent devs from passing wrong type and this can be done at the project level not impacting all the devs around the world.

one detail:
To get the right type for case1 from example on top i had to:
`let spy: CustomSpyInstance<HeroService['computeMaxLevel']>;`
In other case i had a type error when i use the `spyOn` i couldnt solve without this trick.

Comment (nasreddineskandrani) [reply types/jest/index.d.ts]: noImplicitAny didn't work.

if default values (any) used in jest types:
I found a way by wrapping the type
```type CustomSpyInstance<A = never, B = never> = jest.SpyInstance<A, B>;``` which override any
with this it will prevent devs from passing wrong type and this can be done at the project level not impacting all the devs around the world (migrating from js to ts ;) )

**Comment by antoinebrault** [comment]
Added back default values. 
I still want the PR to be merged even thought I'm not in favor of keeping defaults.

@JoshuaKGoldberg 
> Large codebases converting from JavaScript would want as few places to need to explicitly type variables as possible.
What's the point of explicitly declaring jest.SpyInstance and jest.Mock if your not gonna use types. If people don't want to explicitly type variables in tests, they can still use implicit any. IDEs are still gonna autocomplete jest methods.

```typescript
let spy; 
spy = spyOn(obj, 'method');
expect(spy).toHaveBeenCalled();

let foo;
test('foo', () => {
  foo.mockName('MyMock');
  const rendered = render(<Component foo={foo} />);
  expect(rendered).toMatchSnapshot();
});
```

> In a lot of folks' minds, unit tests don't need nearly the same level of type checking as source code, and if you explicitly fail/error within a few milliseconds, needing to write & update them constantly just gets in the way.
This is wrong on so many levels. You want your tests to fail ASAP.
Once you add a new person on a project, the knowledge of that person will be limited. Tests should break when a contract changes (e.g. if an interface/method/property is updated).

Like @nasreddineskandrani said, there will be inconsistencies. Strong typing will be infered from fn(), spyOn(), Mocked<>, but SpyInstance, Mock declarations will not.

Anyway, for future reference, people can use a custom types and ban jest.SpyInstance/jestMock from their codebase if they want be type safe.

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Thanks for easing back on the breaking changes - I actually agree with your points on explicit typing being preferable, but I think we should discuss those breaking changes in a separate issue.

There's still one part I can find that's a breaking change here. This used to be allowed:

```typescript
let spyUntyped: jest.SpyInstance;
spyUntyped = jest.spyOn(spiedTarget, "setValue");
```

...but now it gives a `Type 'SpyInstance<void, [number]>' is not assignable to type 'SpyInstance<{}, any>'`. 

The only simple way to resolve this is to change `SpyInstance` to `<T = any` instead of `<T = {}`. That might itself be a breaking change (at the very least it's a little inaccurate, right) so maybe this isn't something that can be worked around...? Thoughts?

**Comment by antoinebrault** [comment]
@JoshuaKGoldberg you're right. I updated the code and added a test for that.

**Review by JoshuaKGoldberg — APPROVED** [review]
Awesome, thanks!

**Comment by nasreddineskandrani** [comment]
for your attention:
https://github.com/Microsoft/TypeScript/issues/28922

File: types/jest/index.d.ts [hunk 4]
```diff
@@ -110,19 +113,15 @@ declare namespace jest {
     /**
      * Creates a mock function. Optionally takes a mock implementation.
      */
-    function fn<T extends {}>(implementation: (...args: any[]) => T): Mock<T>;
-    /**
-     * Creates a mock function. Optionally takes a mock implementation.
-     */
-    function fn<T>(implementation?: (...args: any[]) => any): Mock<T>;
+    function fn<T, Y extends any[]>(implementation?: (...args: Y) => T): Mock<T, Y>;
```
Comment (Kovensky) [inline types/jest/index.d.ts]: You need `Y extends never[]`, otherwise this refuses functions that have `never` parameters (rare, but possible).

```ts
declare function unreachable(arg: never): never

jest.fn(unreachable) // will error with `any[]`
```

File: types/jest/index.d.ts (hunk 3 above)
Comment (Kovensky) [reply types/jest/index.d.ts]: `extends {}` is already the default; should be `extends object`.

File: types/jest/index.d.ts [hunk 5]
```diff
@@ -198,7 +197,9 @@ declare namespace jest {
      *   spy.mockRestore();
      * });
      */
-    function spyOn<T extends {}, M extends keyof T>(object: T, method: M, accessType?: 'get' | 'set'): SpyInstance<T[M]>;
+    function spyOn<T extends {}, M extends keyof T>(object: T, method: M, accessType: 'get'): SpyInstance<T[M], []>;
+    function spyOn<T extends {}, M extends keyof T>(object: T, method: M, accessType: 'set'): SpyInstance<void, [T[M]]>;
+    function spyOn<T extends {}, M extends keyof T>(object: T, method: M): T[M] extends (...args: any[]) => any ? SpyInstance<ReturnType<T[M]>, ArgsType<T[M]>> : never;
```
Comment (Kovensky) [inline types/jest/index.d.ts]: `extends (...args: never[]) => unknown`  is the most compatible function signature

File: types/jest/index.d.ts [hunk 6]
```diff
@@ -762,12 +763,12 @@ declare namespace jest {
         new (...args: any[]): any;
     }
 
-    interface Mock<T = {}> extends Function, MockInstance<T> {
-        new (...args: any[]): T;
-        (...args: any[]): any;
+    interface Mock<T = any, Y extends any[] = any> extends Function, MockInstance<T, Y> {
```
Comment (Kovensky) [inline types/jest/index.d.ts]: `extends never[] = any[]`

File: types/jest/index.d.ts [hunk 7]
```diff
@@ -762,12 +763,12 @@ declare namespace jest {
...
+        new (...args: Y): T;
+        (...args: Y): T;
     }
 
-    interface SpyInstance<T = {}> extends MockInstance<T> {}
+    interface SpyInstance<T = any, Y extends any[] = any> extends MockInstance<T, Y> {}
```
Comment (Kovensky) [inline types/jest/index.d.ts]: Same here

File: types/jest/index.d.ts (hunk 4 above)
Comment (Kovensky) [reply types/jest/index.d.ts]: Actually, is there any reason why this is not just `function fn<T extends (...args: never[]) => unknown>(implementation?: T): Mock<T>`, with `Mock` changed to hold a function generic instead?

It makes declaring the `Mock` methods more annoying _but_ it makes using `.fn` a lot easier when you want to make sure your mock function conforms to the interface of a function that already exists. Just call `jest.fn<typeof myfunc>(/* inline function will even get contextual params/return types */)`.

Comment (Kovensky) [reply types/jest/index.d.ts]: Also, the default type (when implementation is not specified) should be `(...args: any[]) => void`. Perhaps it should be on a 0 argument overload instead of an optional argument to avoid misuse.

**Review by Jessidhia — CHANGES_REQUESTED** [review]
~~Yeeeees, I've been yearning for this~~

The idea is great, but I think `jest.fn` can be made even easier to use by receiving a function type itself as the generic argument instead of separate args/return.

File: types/jest/index.d.ts (hunk 5 above)
Comment (Kovensky) [reply types/jest/index.d.ts]: This `never` return sucks but there's really no better way, huh... 😢 

`<..., T[M] extends (...args: never[]) => unknown>` would DWIM but 1) it's a syntax error; 2) it's not a real parameter, just an additional constraint, and it can't even be named.

File: types/jest/index.d.ts (hunk 4 above)
Comment (antoinebrault) [reply types/jest/index.d.ts]: > You need `Y extends never[]`, otherwise this refuses functions that have `never` parameters (rare, but possible).
> 
> ```ts
> declare function unreachable(arg: never): never
> 
> jest.fn(unreachable) // will error with `any[]`
> ```
This is false. I just updated the tests and it doesn't error. 



> Actually, is there any reason why this is not just `function fn<T extends (...args: never[]) => unknown>(implementation?: T): Mock<T>`, with `Mock` changed to hold a function generic instead?
> 
> It makes declaring the `Mock` methods more annoying _but_ it makes using `.fn` a lot easier when you want to make sure your mock function conforms to the interface of a function that already exists. Just call `jest.fn<typeof myfunc>(/* inline function will even get contextual params/return types */)`.

Because this would be a breaking change. By adding a new generic with a default value, existing code will not be affected.



> Also, the default type (when implementation is not specified) should be `(...args: any[]) => void`. Perhaps it should be on a 0 argument overload instead of an optional argument to avoid misuse.

It should not return void. This should compile:
```
jest.fn().mockImplementation((test: number) => test);
```

Comment (Jessidhia) [reply types/jest/index.d.ts]: Hmm... I need to brush up on the `...args: never[]` thing, then. I had seen it as a recommendation from the TS team *somewhere* a few months ago.

As for the second one, that's because you're specifying a mock implementation after the `.fn` call but... that is supported API because JS is weird like that 😵

If you use `.fn()` _without_ also adding a more specific `.mock*` it'll return `void` by default. But currently methods cannot alter the type of anything on their `this`...

Comment (antoinebrault) [reply types/jest/index.d.ts]: FYI opened a new PR #32579

Comment (josecolella) [reply types/jest/index.d.ts]: @antoinebrault Can you help me understand how to use the new:

```ts
function fn<T, Y extends any[]>(implementation: (...args: Y) => T): Mock<T, Y>;
```

I understand the T, but what is expected for the Y. After updating to this version, several of the mockImplementations, are throwing typing issues.

Comment (antoinebrault) [reply types/jest/index.d.ts]: @josecolella Y is the arguments types of the implementation function.

e.g.
```typescript
function x(y: number) {
  return y + '';
}
jest.fn(x) // will be Mock<string, [number]>
```

When mocking implementations, you have to use the same arguments types as the original function, but arguments are optional.
e.g. 
```typescript
function x(y: number) {
  return y + '';
}
jest.fn(x).mockImplementation(() => ''); // will work
```

Comment (josecolella) [reply types/jest/index.d.ts]: @antoinebrault What about for mockImplementation of classes? As of now it's complaining that I need to implement all the methods within the class, when in reality only one method needs to be mocked. 

```ts
interface SomeInterface {
      foo() 
      foo2()
}

jest.fn<SomeInterface>(() => ({
 foo: jest.fn().mockReturnValue(true)
}));
```

With new types this no longer works. Do you have an example of how to use this new version with class mock implementations?

Comment (antoinebrault) [reply types/jest/index.d.ts]: > @antoinebrault What about for mockImplementation of classes? As of now it's complaining that I need to implement all the methods within the class, when in reality only one method needs to be mocked.
> 
you can use Partial for that usecase.  ```jest.Mocked<Partial<Type>>```

> ```ts
> interface SomeInterface {
>       foo() 
>       foo2()
> }
> 
> jest.fn<SomeInterface>(() => ({
>  foo: jest.fn().mockReturnValue(true)
> }));
> ```
> 
> With new types this no longer works. Do you have an example of how to use this new version with class mock implementations?

```typescript
interface SomeInterface {
    foo(): string,
    foo2(): number
}
const y: jest.Mocked<SomeInterface> = {
    foo: jest.fn().mockReturnValue(''),
    foo2: jest.fn().mockReturnValue(1)
};
```

Comment (josecolella) [reply types/jest/index.d.ts]: @antoinebrault Thanks for your explanation and examples.
